torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -95,16 +97,58 @@ class FullyShardedDataParallel(nn.Module):
             params and grads to be on same device to work with optimizer. This
             API is subject to change. Default is ``None`` in which case there
             will be no offloading.
+        auto_wrap_policy: (Optional [callable]):
+            A callable specifying a policy to recursively wrap layers with FSDP.
+            Note that this policy currently will only apply to child modules of
+            the passed in module. The overall module is always FSDP because we
+            call the constructor.
     """
 
     def __init__(
         self,
         module: nn.Module,
         process_group: Optional[ProcessGroup] = None,
-        cpu_offload: Optional[CPUOffload] = None
+        cpu_offload: Optional[CPUOffload] = None,
+        fsdp_auto_wrap_policy: Optional[callable] = None,
     ):
         torch._C._log_api_usage_once("torch.distributed.fsdp")
         super().__init__()
+        # if fsdp_auto_wrap_policy is specified, submodules should not be
+        # already wrapped, otherwise we'd attempt to double wrap them resulting
+        # in errors.
+        if fsdp_auto_wrap_policy is not None:
+            num_params = sum([p.numel() for p in module.parameters()])
+            self._check_wrapped(
+                module,
+                check_fn=lambda mod: not isinstance(mod, FullyShardedDataParallel),
+                err_fn=lambda mod: f"Expected {mod} to NOT be FullyShardedDataParallel if auto_wrap is enabled.",
+            )
+            # TODO: refactor recursive_wrap so that it is not dependent on
+            # ConfigAutoWrap.
```

Comment:
agree with this, we need to refactor this in the follow up PRs

Reply:
Yep, will remove it in follow up PRs.